mage_integrations/mage_integrations/destinations/snowflake/__init__.py

```diff
@@ -334,6 +317,97 @@ def calculate_records_inserted_and_updated(
+                record_data,
+                stream,
+                tags,
+                **kwargs,
+            )
+        else:
+            results = []
+            self.logger.info('Using batch load for Snowflake...')
+
+            unique_constraints = self.unique_constraints.get(stream)
+            unique_conflict_method = self.unique_conflict_methods.get(stream)
+            self.logger.info(f'Batch upload unique_constraints: {unique_constraints}')
+            self.logger.info(f'Batch upload unique_conflict_method: {unique_conflict_method}')
+
+            # Execute the query_strings before inserting the data, i.e., passed-in
+            # query_strings including create table command or alter table command
+            if query_strings and len(query_strings) > 0:
+                self.logger.info(f'Executing query_strings: {query_strings}')
+                results += self.build_connection().execute(query_strings, commit=True)
+            else:
+                self.logger.info(f'Skip executing empty query_strings: {query_strings}')
+
+            df = pd.DataFrame([d['record'] for d in record_data])
+            df.columns = df.columns.str.lower()
+            self.logger.info(f'Batch upload to Snowflake: {df.shape[0]} rows.')
+            self.logger.info(f'Columns: {df.columns}')
+            database = self.config.get(self.DATABASE_CONFIG_KEY)
+            schema = self.config.get(self.SCHEMA_CONFIG_KEY)
+            table = self.config.get(self.TABLE_CONFIG_KEY)
+
+            if unique_constraints and unique_conflict_method:
+                full_table_name = self.full_table_name(database, schema, table)
+                full_table_name_temp = self.full_table_name_temp(database, schema, table)
+                drop_temp_table_command = [f'DROP TABLE IF EXISTS {full_table_name_temp}']
+
+                results += self.build_connection().execute(drop_temp_table_command, commit=True)
+
+                # Outputs of write_dataframe_to_table are for temporary table only, thus not added to results
+                # results += self.write_dataframe_to_table(df, database, schema, f'temp_{table}')
+                self.write_dataframe_to_table(df, database, schema, f'temp_{table}')
+
+                merge_command = [self.build_merge_command(
+                    columns=df.columns,
+                    full_table_name=full_table_name,
+                    full_table_name_temp=full_table_name_temp,
+                    unique_conflict_method=unique_conflict_method,
+                    unique_constraints=unique_constraints,
+                )]
+                self.logger.info(f'Merging {full_table_name_temp} into {full_table_name}')
+                results += self.build_connection().execute(merge_command, commit=True)
+
+                self.logger.info(f'Dropping temporary table: {full_table_name_temp}')
+                results += self.build_connection().execute(drop_temp_table_command, commit=True)
```

Comment:
The merge_command and drop_temp_table_command can be executed in the same snowflake connection.

Reply:
Good catch: made the suggested change.